File: problems/HackerRank/2521_exercises/tree/getMinimumDifference.c
int min(int a, int b)
{
	return a < b ? a : b;
}

// use dfs in order
void dfs(struct TreeNode *root, int *prev, int *minValue)
{

	if (root == NULL)
	{
		return;
	}

	dfs(root->left, prev, minValue);

	if (*prev != -1)
	{
		int diff = abs(*prev - root->val);
		*minValue = min(diff, *minValue);
	}

	*prev = root->val;

	dfs(root->right, prev, minValue);
}

int getMinimumDifference(struct TreeNode *root)
{
	int minValue = INT_MAX;
	int prev = -1;
	dfs(root, &prev, &minValue);
	return minValue;
}

File: problems/LeetCode/TopInterview150/getMinimumDifference.ts
function getMinimumDifference(root: TreeNode | null): number {
  let min = Infinity;
  let pre = -Infinity;

  // inorder traverse
  const dfs = (root: TreeNode | null) => {
    if (!root) return;

    dfs(root.left);
    min = Math.min(min, root.val - pre);
    pre = root.val;
    dfs(root.right);
  };

  dfs(root);

  return min;
}

// Time compleixty: O(n)
// Space complexity: O(h)
